src/Lykke.AzureStorage/Tables/Decorators/ExplicitAppInsightsAzureTableStorageDecorator.cs

```diff
@@ -0,0 +1,262 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using System.Diagnostics;
+using Microsoft.ApplicationInsights;
+using Microsoft.WindowsAzure.Storage.Table;
+using Lykke.AzureStorage.Tables.Paging;
+
+namespace AzureStorage.Tables.Decorators
+{
+    /// <summary>
+    /// Decorator which explicitly calls AppInsights API to submit Azure Table Storage call events
+    /// </summary>
+    internal class ExplicitAppInsightsAzureTableStorageDecorator<TEntity> : INoSQLTableStorage<TEntity>
+        where TEntity : ITableEntity, new()
+    {
+        private const string _telemetryDependencyType = "Azure table";
+        private readonly INoSQLTableStorage<TEntity> _impl;
+        private readonly TelemetryClient _telemetry = new TelemetryClient();
+
+        public string Name => _impl.Name;
+
+        internal ExplicitAppInsightsAzureTableStorageDecorator(INoSQLTableStorage<TEntity> storage)
+        {
+            _impl = storage;
+        }
+
+        #region INoSqlTableStorage{TEntity} decoration
+
+        public IEnumerator<TEntity> GetEnumerator()
+            => Wrap(() => _impl.GetEnumerator(), nameof(GetEnumerator));
+
+        IEnumerator IEnumerable.GetEnumerator()
+            => Wrap(() => ((IEnumerable)_impl).GetEnumerator(), nameof(GetEnumerator));
+
+        TEntity INoSQLTableStorage<TEntity>.this[string partition, string row]
+            => Wrap(() => _impl[partition, row], "[partition, row]");
+
+        IEnumerable<TEntity> INoSQLTableStorage<TEntity>.this[string partition]
+            => Wrap(() => _impl[partition], "[partition]");
+
+        public Task InsertAsync(TEntity item, params int[] notLogCodes)
+            => WrapAsync(() => _impl.InsertAsync(item, notLogCodes), nameof(InsertAsync));
+
+        public Task InsertAsync(IEnumerable<TEntity> items)
+            => WrapAsync(() => _impl.InsertAsync(items), nameof(InsertAsync));
+
+        public Task InsertOrMergeAsync(TEntity item)
+            => WrapAsync(() => _impl.InsertOrMergeAsync(item), nameof(InsertOrMergeAsync));
+
+        public Task InsertOrMergeBatchAsync(IEnumerable<TEntity> items)
+            => WrapAsync(() => _impl.InsertOrMergeBatchAsync(items), nameof(InsertOrMergeBatchAsync));
+
+        public Task ReplaceAsync(TEntity entity)
+            => WrapAsync(() => _impl.ReplaceAsync(entity), nameof(ReplaceAsync));
```

Comment:
I think for the methods that have multiple overloads, nameof(XXX) is not enough. I think something like GetMethodSignature() should be added and used everywhere. This method could return "ReplaceAsync(TEntity)" and "ReplaceAsync(string,string,Func<TEntity,TEntuity>" for this particular case